File: lib-rtmp/src/main/java/com/hezb/lib/rtmp/RtmpClient.java
package com.hezb.lib.rtmp;

/**
 * Project Name: AndroidScreenLive
 * File Name:    RtmpClient
 *
 * Description: rtmp客户端.
 *
 * @author  hezhubo
 * @date    2022年07月21日 13:00
 */
public class RtmpClient {

    static {
        System.loadLibrary("rtmp");
    }

    private long handle = 0; // 底层实例指针

    public native static String getVersion();

    private native long open(String url, boolean isPublishMode);

    private native int read(long rtmpPointer, byte[] data, int offset, int size);

    private native int write(long rtmpPointer, byte[] data, int size, int type, int ts);

    private native void close(long rtmpPointer);

    public boolean isConnected() {
        return handle != 0;
    }

    public boolean connect(String url, boolean isPublishMode) {
        close();
        // 底层返回rtmp的指针; handle == 0 则创建失败
        handle = open(url, isPublishMode);
        return handle != 0;
    }

    public int read(byte[] data, int offset, int size) {
        if (handle != 0) {
            return read(handle, data, offset, size);
        }
        return 0;
    }

    public boolean write(byte[] data, int size, int type, int ts) {
        if (handle != 0) {
            return write(handle, data, size, type, ts) == 1;
        }
        return false;
    }

    public void close() {
        if (handle != 0) {
            close(handle);
            handle = 0;
        }
    }

}

File: app/src/main/java/com/hezb/live/recorder/rtmp/RtmpClient.kt
package com.hezb.live.recorder.rtmp

import androidx.annotation.Keep
import com.hezb.live.recorder.util.LogUtil

/**
 * Project Name: AndroidScreenLive
 * File Name:    RtmpClient
 *
 * Description: rtmp客户端.
 *
 * @author  hezhubo
 * @date    2022年07月21日 13:00
 */
@Keep
object RtmpClient {

    init {
        try {
            System.loadLibrary("rtmp")
        } catch (t: Throwable) {
            LogUtil.e(msg = "load rtmp so error!", tr = t)
        }
    }

    external fun open(url: String, isPublishMode: Boolean): Long

    external fun read(rtmpPointer: Long, data: ByteArray, offset: Int, size: Int): Int

    external fun write(rtmpPointer: Long, data: ByteArray, size: Int, type: Int, ts: Int): Int

    external fun close(rtmpPointer: Long): Int

    external fun getIpAddr(rtmpPointer: Long): String?

}